[WIP] Add S4 decoder in ESPnet2

m-koichi opened the pull request:
Hi, this PR adds an implementation of the S4 decoder (https://arxiv.org/abs/2210.17098) and a sample config.
The S4 codebase follows the official repository (https://github.com/HazyResearch/state-spaces), so the network configuration format is somewhat different from the models supported in ESPnet (e.g., Transformer decoder).

- LibriSpeech 960h WER result using this PR

|                     | dev_clean | dev_other | test_clean | test_other |
|:-----------:|----------:|-----------:|-----------:|-----------:|
| without LM | 2.0       | 5.0       | 2.3        | 5.0       |
| with LM       | 1.7       | 4.0       | 1.9        | 4.1        |

sw005320 commented:
Many thanks!
@pyf98, can you review this pr?

sw005320 commented on egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:
```diff
@@ -0,0 +1,93 @@
+    cnn_module_kernel: 31
+
+decoder: s4
+decoder_conf:
```
Can you add some comments that the network configuration format is different?

sw005320 commented on espnet/nets/pytorch_backend/state_spaces/attention.py:
```diff
@@ -0,0 +1,119 @@
+from espnet.nets.pytorch_backend.state_spaces.base import SequenceModule
+
+
+class MultiHeadedAttention(SequenceModule):
```
Can you add a comment about why you need this instead of using our default MHA?

sw005320 reviewed (commented):
- We may move the state_spaces directory under `espnet2/asr/` (if this is not easy or there is another reason, of course, we can keep it here).
- Can you upload a model checkpoint to the HF hub and add a model link and its result to `librispeech/asr1/REAME.md`?
- Can you also add a test? We can consider it later.

m-koichi commented:
Sorry for the late reply.
> We may move the state_spaces directory under espnet2/asr/ (if this is not easy or there is another reason, of course, we can keep it here).

Since we tested S4 decoder in ASR and TTS tasks, we placed state_spaces directory under espnet/nets/pytorch_backend. Now I have moved the repository under espnet2/asr/
> Can you upload a model checkpoint to the HF hub and add a model link and its result to librispeech/asr1/REAME.md?

I have uploaded a model checkpoint to the HF hub [[link](https://huggingface.co/espnet/kmiyazaki_librispeech_asr_s4_decoder)]. I will update model card README.md later.

> Can you also add a test? We can consider it later.

I have added a simple test.

pyf98 commented on egs2/librispeech/asr1/conf/tuning/train_asr_s4_decoder.yaml:
```diff
@@ -0,0 +1,96 @@
+    - max
+keep_nbest_models: 10
+
+optim: adamw
+exclude_weight_decay: true
```
Just curious: Is AdamW better than Adam? Also, does excluding weight decay for bias etc improve the performance?

m-koichi replied:
In our experiments, in S4 decoder case, we confirmed that AdamW is slightly better than Adam optimizer.
While in Transformer decoder case, there was no significant difference. 
For the weight decay trick, we followed the experimental setup of S4. Perhaps the effect may not be so apparent.

- S4 results (w/o LM)

|                     | dev_clean | dev_other | test_clean | test_other |
|---------------------|-----------|-----------|------------|------------|
| Adam                | 2.1       | 5.1       | 2.2        | 5.2        |
| AdamW               | 2.0       | 5.0       | 2.3        | 5.0        |
| AdamW w/o excluding | 2.0       | 5.0       | 2.2        | 5.0        |

pyf98 replied:
Thanks!

b-flo replied:
Kind of out of scope but did someone experiment with the `ScaledAdam` from k2 team?

pyf98 commented on espnet2/tasks/abs_task.py:
```diff
@@ -817,6 +818,19 @@ class ArgumentDefaultsRawTextHelpFormatter(
+            "--exclude_weight_decay_conf",
+            action=NestedDictAction,
+            default=dict(),
+            help="The keyword arguments for configuring weight decay in optimizer",
```
Can you explain a bit more about this new argument?

m-koichi replied:
Updated. Could you check the comment again?

pyf98 commented:
Overall, it looks good to me.

sw005320 commented:
Thanks, @pyf98 

@m-koichi, after you fix the CI issue and reflect @pyf98's comments, I'll merge this PR.

sw005320 commented:
We still have some CI errors.

sw005320 commented:
https://github.com/espnet/espnet/actions/runs/3881628865/jobs/6637706128#step:8:18610
I think you can skip the test for the lower versions.
You can find some examples (e.g., https://github.com/espnet/espnet/blob/master/test/espnet2/enh/layers/test_complex_utils.py)

m-koichi commented:
Thanks for the advice.
I fixed the test to skip for the lower pytorch versions.

sw005320 commented:
Many many thanks!
This is a really valuable contribution!

Could you add this new feature to https://github.com/espnet/espnet/blob/master/README.md to advertise this great function?

m-koichi commented:
Thank you for reviewing the code and helping some CI issues.
All comments received have been reflected.
I updated HF hub model card to reproduce the result.
I added a bullet to the README related to the decoders of the ASR architecture.
`- Decoder: RNN (LSTM/GRU), Transformer, or S4`
Can I continue to commit to this branch? or should I create another pull request?

sw005320 commented:
You can create another PR.

ftshijt commented:
Thanks for the contribution! 

I found that S4decoder needs an additional `hydra` package imported in `espnet2/asr/state_spaces/utils.py`. After my merge to this PR, it throws an error about `No module named 'hydra'` where I need to pip install. Not sure if we need to either check the import as an external package (as in the installer or makefile) or add that package to the `setup.py`
